layers/shader_object.cpp

```diff
@@ -0,0 +1,4044 @@
+    return cmd_data;
+}
+
+void CommandBufferData::Destroy(CommandBufferData** data) {
+    ASSERT(data);
+    VkAllocationCallbacks allocator = (*data)->allocator;
+    allocator.pfnFree(allocator.pUserData, *data);
+    *data = nullptr;
+}
+
+VkResult Shader::Create(DeviceData const& deviceData, VkShaderCreateInfoEXT const& createInfo, VkAllocationCallbacks const& allocator,
+                        Shader** ppOutShader) {
+    auto& vtable = deviceData.vtable;
+
+    // Get SPIR-V information
+    size_t spirv_size;
+    void const* spirv_data;
+    if (createInfo.codeType == VK_SHADER_CODE_TYPE_SPIRV_EXT) {
+        // SPIR-V is stored directly in the create info
+        spirv_size = createInfo.codeSize;
+        spirv_data = createInfo.pCode;
+    } else if (createInfo.codeType == VK_SHADER_CODE_TYPE_BINARY_EXT && ContainsValidShaderBinary(deviceData, createInfo)) {
+        // SPIR-V is stored in the shader binary
+        auto shader_binary = static_cast<ShaderBinary const*>(createInfo.pCode);
+        spirv_size = shader_binary->spirv_data_size;
+        spirv_data = shader_binary->GetSprivData();
+    } else {
+        return VK_ERROR_INCOMPATIBLE_SHADER_BINARY_EXT;
+    }
+
+    size_t name_size = createInfo.pName == nullptr ? 0 : strlen(createInfo.pName) + 1;
+
+    size_t bytes_to_allocate = sizeof(Shader);
+    bytes_to_allocate += name_size;
+    bytes_to_allocate += spirv_size;
+    bytes_to_allocate += createInfo.pushConstantRangeCount           * sizeof(*createInfo.pPushConstantRanges);
+    bytes_to_allocate += createInfo.setLayoutCount                   * sizeof(*createInfo.pSetLayouts);
+    bytes_to_allocate += deviceData.reserved_private_data_slot_count * sizeof(PrivateDataSlotPair);
+    if (createInfo.pSpecializationInfo) {
+        bytes_to_allocate += sizeof(*createInfo.pSpecializationInfo);
+        bytes_to_allocate += createInfo.pSpecializationInfo->dataSize;
+        bytes_to_allocate += createInfo.pSpecializationInfo->mapEntryCount * sizeof(VkSpecializationMapEntry);
+    }
+
+    void* memory = allocator.pfnAllocation(allocator.pUserData, bytes_to_allocate, 0,
+                                           VkSystemAllocationScope::VK_SYSTEM_ALLOCATION_SCOPE_OBJECT);
+    if (!memory) {
+        return VK_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    Shader* shader = new (memory) Shader();
+    *ppOutShader = shader;
+    shader->stage = createInfo.stage;
+    if (createInfo.flags & VK_SHADER_CREATE_ALLOW_VARYING_SUBGROUP_SIZE_BIT_EXT) {
+        shader->flags |= VK_PIPELINE_SHADER_STAGE_CREATE_ALLOW_VARYING_SUBGROUP_SIZE_BIT;
+    }
+    if (createInfo.flags & VK_SHADER_CREATE_REQUIRE_FULL_SUBGROUPS_BIT_EXT) {
+        shader->flags |= VK_PIPELINE_SHADER_STAGE_CREATE_REQUIRE_FULL_SUBGROUPS_BIT;
+    }
+
+    // Copy data over from create info struct
+
+    uint8_t* offset = reinterpret_cast<uint8_t*>(shader) + sizeof(Shader);
+
+#define COPY_BYTES(dst_ptr, dst_size, src_ptr, src_size) \
```

Comment:
I'm not a huge fan of these macros defined inside a single function to copy a single structure type.  Could this be done instead with templated helper functions?